README.md
# Building Office Add-ins using Office.js

## The everygreen book

This repo is a sample-code and issue-tracking companion to the book ["Building Office Add-ins using Office.js"](https://leanpub.com/buildingofficeaddins/) -- an "evergreen" e-book published and purchasable through [LeanPub.com](https://leanpub.com/buildingofficeaddins/).

From the book description:

> This book is a guide to Office.js APIs, which are used to build the web-technology-based Office Add-ins. Is describes the principles and design patterns shared by the new 2016+ wave of APIs for Excel, Word, and OneNote -- written by someone who, as member of the Office Extensibility Platform team at Microsoft, had helped design them. 

&nbsp;

## This repo's purpose is to:

**\#1: Allow me to share the larger code samples and associated assets** (e.g., documents with sample data, geared specifically for the scenario).

**\#2: Provides readers with a way to log and track content issues (and be informed once they are fixed).  Issues can be:**

  * Simple content issues: A misspelling, an incomplete phrase, a sentence that no longer makes sense.
  * Requests for additional detail in *existing* topics.
  * Requests for *brand new topics*. I might already be planning to write something about it eventually, but this will let you subscribe to finding out when the content is made available; and will also help me gauge topic interest.
  * Issues with sample code or related assets: code is unclear; the design is sub optimal; something needs a comment; or perhaps an existing comment needs to be clarified or removed.
  * Anything else?

**\#3: Keep track of the [Release notes](Release-Notes.md)**.

&nbsp;


## If you have questions about specific APIs, please use StackOverflow.

This repo is meant for aggregating book issues -- but the book's purpose is *not* to cover every single API, but rather to cover the common concepts.  If you have a question about a specific API, please use [**StackOverflow**, tagging your question **office-js**](stackoverflow.com/questions/tagged/office-js).

In fact, even if your question *is* about a core concept, I still encourage you to ask it on StackOverflow first, as you'll almost certainly get a quicker response there (and if the response is from me, it might well make it into the book eventually!).  This repo is for *tracking* the issues and topics, but addressing questions is best done through [**StackOverflow**](stackoverflow.com/questions/tagged/office-js).


If you want to contact me, please feel free to reach out at

![](http://buildingofficeaddins.com/wp-content/uploads/michael-at-buildingofficeaddins.png)


Release-Notes.md
## Book release notes

Updates to the book are free for all existing readers (which is what lean-publishing, and the concept of an evergreen book, is all about!).  Simply go to **<https://leanpub.com/user_dashboard/library>**, select the book, and download your newly-updated copy!

For any issues or topic requests, please file an issue on <http://buildingofficeaddins.com/issues>.

&nbsp;

### Version 1.6 (*August 26, 2017*) [261 pages]

* Major re-structuring of the book, splitting out the former chapters 5 & 6 -- which were bursting at the seams -- into a bunch of smaller chapters.  Also streamlined the rest of the book, moving topics that were less immediately-necessary (e.g., API versioning) further towards the back of the book, to make it faster to get started.
* As part of the getting-started chapter, added a section for my recommendations of "*The optimal dev environment*"
* Expanded the section on "*Handling errors*.
* Added "*Recap: the four basic principles of Office.js*" to the chapter on core Office.js topics.
* Added runnable Script Lab snippets to "*Canonical code sample: reading data and performing actions on the document*" (including a refactored version that splits the task into multiple subroutines, and a plain ES5 JavaScript variant).

&nbsp;

### Version 1.5 (*August 2, 2017*) [249 pages]

* By popular demand, added a massively-detailed (15+ pages) and example-filled section on "*Using objects outside the "linear" `Excel.run` or `Word.run` flow (e.g., in a button-click callback, in a `setInterval`, etc.)*".  Its subsections include:
  * Re-hydrating an existing Request Context: the overall pattern, proper error-handling, `object.track`, and cleanup of tracked objects]
  * A common, and infuriatingly silent, mistake:  queueing up actions on the wrong request context
  * Resuming with multiple objects
  * Why can't we have a single global request context, and be one happy family?
* Addressed a couple of [small reader-reported issues](https://github.com/Zlatkovsky/BuildingOfficeAddins/milestone/8?closed=1).

&nbsp;

### Version 1.4 (*July 19, 2017*) [232 pages]

* Added a detailed section for how to check whether an object exists (e.g., whether there's a worksheet by a particular name, whether two ranges intersect, etc). The section also talks about the powerful but unusual "null object" pattern, used in methods and properties suffixed with `*OrNullObject`.
* Added a section with links to "*API documentation resources*".
* Addressed a couple of [small reader-reported issues](https://github.com/Zlatkovsky/BuildingOfficeAddins/milestone/7?closed=1).

&nbsp;

### Version 1.3 (*May 23, 2017*)
 
* Added an entire section devoted to the `PropertyNotLoaded` error.
* Added info about [Script Lab](https://aka.ms/scriptlab), a playground tool to easily try out code samples described in this book
* Addressed a variety of [small reader-reported issues](https://github.com/Zlatkovsky/BuildingOfficeAddins/milestone/5?closed=1).

&nbsp;

### Version 1.2  (*Feb 20, 2017*)

* Added a topic on the different flavors of Office 2016 / Office 365 -- and the practical implications for developers.
* Added a topic on API Versioning and Requirement Sets.
* Greatly expanded the "TypeScript-based Add-ins" topic, adding instructions for the updated Yeoman generator.
* Added a topic for attaching the debugger to Add-ins (breakpoints, DOM explorer, etc.)
* Added a link to the book's companion Twitter account.
* Addressed a number of other [reader-reported issues](https://github.com/Zlatkovsky/BuildingOfficeAddins/milestone/4?closed=1).

&nbsp;

### Version 1.1  (*Jan 22, 2017*)

* Re-pivoted the book around TypeScript and the `async/await` syntax. Moved the JS-specific content to a separate Appendix
* Added an information-packed JavaScript & TypeScript crash-course, tailored specifically at Office.js concepts, for those who are new to the world of JS/TS.
* With TypeScript now a first-class citizen of the book, added "*Getting started with building TypeScript-based Add-ins*" (section 3.2).  I expect to continue to expand this section in future releases.
* Added an in-depth explanation of the internal workings of the Office.js pipeline and proxy-object model.  See "*Implementation details, for those who want to know how it really works*" (section 5.5).
* Re-arranged, edited, and added to the content of the "*Office.js APIs: Core concepts*" chapter (chapter 5).
* Added links to downloadable code samples, for a few of the larger samples.
* Addressed a variety of smaller feedback items.

&nbsp;

### Updates are free for existing readers

Updates to the book are free for all existing readers (which is what lean-publishing, and the concept of an evergreen book, is all about!).  Simply go to <https://leanpub.com/user_dashboard/library>, select the book, and download your newly-updated copy!

If you haven't purchased the book yet, you can do so at <https://leanpub.com/buildingofficeaddins>.


Samples/Chapter 5 - Core Concepts/Stocks/Refresh stocks.ts
function refreshStocks() {
    Excel.run(async (context) => {
        console.log("#1: Obtain and loading worksheets");
        let sheets = context.workbook.worksheets.load("name");
        await context.sync();

        console.log("#2: Retrieve stock table & values. " +
            "Also load tables on previous sheet, if any");
        let latestSheet = sheets.items[sheets.items.length - 1];
        let latestTable = latestSheet.tables.getItemAt(0);
        let stocksRange = latestTable.getDataBodyRange()
            .getColumn(0).load("values");
        let previousTables = queueLoadPreviousSheetTables(sheets);
        await context.sync();

        console.log("#3: Issue web request to read stock data");
        let stockNames = stocksRange.values.map(rowData => rowData[0]);
        
        // With this names-list in hand, call the "getStockData"
        // function from an earlier chapter ("Promises Primer"").
        // The function returns a Promise, so await it:
        let stockData = await getStockData(stockNames);
        console.log("Received stock data", stockData);

        console.log("#4: Reload stock names, to ensure no " +
            "changes occurred while waiting on the web request");
        stocksRange.load("values");
        await context.sync();

        console.log("#5: Write in the updated stock prices, " + 
            "and load the new total (and previous, if any)");
        queueUpdateTable(latestTable, stocksRange, stockData);
        let latestTotalCell = latestTable
            .getTotalRowRange().getLastCell().load("values");
        let previousTotalCellIfAny =
            queueLoadPreviousTotalCellIfAny(previousTables);
        await context.sync();

        console.log("#6: If there was a previous 'Total' to " +
            "compare against, color the current one accordingly.");
        queueCellHighlightingIfAny(
            latestTotalCell, previousTotalCellIfAny);
        await context.sync();
        
        console.log(`Done! Data on sheet "${latestSheet.name}" ` +
            "has been refreshed");

    }).catch(OfficeHelpers.Utilities.log);


    // Helper functions

    function queueLoadPreviousSheetTables(
        sheets: Excel.WorksheetCollection
    ) {
        let sheetCount = sheets.items.length;
        if (sheetCount >= 2) {
            return sheets.items[sheetCount - 2].tables.load("name");
        }
        return null;
    }

    function queueUpdateTable(table: Excel.Table,
        stocksRange: Excel.Range, data: any
    ) {
        let pricesToWrite = stocksRange.values.map(row => {
            let stockName = row[0];
            let priceOrEmptyString = data[stockName];
            if (typeof priceOrEmptyString === "undefined") {
                priceOrEmptyString = "";
            }
            return [priceOrEmptyString];
        });

        let priceColumn = table.columns.getItem("Price")
            .getRange().getIntersection(stocksRange.getEntireRow());
        priceColumn.values = pricesToWrite;
    }

    function queueLoadPreviousTotalCellIfAny(
        tables: Excel.TableCollection
    ) {
        if (tables && tables.items.length === 1) {
            return tables.items[0].getTotalRowRange()
                .getLastCell().load("values");
        }
        return null;
    }

    function queueCellHighlightingIfAny(
        latestCell: Excel.Range, previousCell: Excel.Range
    ) {
        if (previousCell) {
            let isLatestGreater =
                latestCell.values[0][0] >
                previousCell.values[0][0];

            latestCell.format.fill.color =
                (isLatestGreater ? "#82E0AA" : "#EC7063");     
        } else {
            console.log("Skipped comparison with previous " +
                "total, as there doesn't appear to be one");
        }
    }
}


function getStockData(stockNamesList: string[]) {
    var quotedCommaSeparatedNames = stockNamesList
        .map(function(name) {
            return '"' + name + '"'
        })
        .join(",");

    var url = '//query.yahooapis.com/v1/public/yql';
    var data = 'q=' +
        encodeURIComponent(
            'select * from yahoo.finance.quotes ' +
            'where symbol in (' + quotedCommaSeparatedNames + ')'
        ) +
        "&env=http%3A%2F%2Fdatatables.org" +
        "%2Falltables.env&format=json";

    return new OfficeExtension.Promise((resolve, reject) => {
        $.ajax({
            url: url,
            dataType: 'json',
            data: data,
            timeout: 5000
        })
        .done((result) => {
            console.log("Web request succeeded");
            var stockDataDictionary = {};
            var stockDataArray = result.query.results.quote;
            stockDataArray.forEach(data => {
                var name = data['Symbol']
                var price = data['LastTradePriceOnly'];
                stockDataDictionary[name] = price;
            });

            resolve(stockDataDictionary);
        })
        .fail((error)=> {
            console.log("Web request failed:");
            console.log(url);

            reject(new Error(error.statusText));
        });
    });
}


Samples/Chapter 5 - Core Concepts/Gradebook/Filter student grades (JavaScript).js
function filterStudentGrades() {
    Excel.run(function(context) {
        // #1: Find the row that matches the name of the student:

        var sheet = context.workbook.worksheets.getActiveWorksheet();
        var nameColumn = sheet.getRange("A:A");
        
        var studentName = $("#student-name").val();
        var matchingRowNum = context.workbook.functions.match(
            studentName, nameColumn, 0 /*exact match*/);
        matchingRowNum.load("value");

        var studentRow; // declared here for passing between "`.then`"s
        
        return context.sync()
            .then(function() {
                // #2: Load the cell values (filtered to just the
                //     used range, to minimize data-transfer)

                studentRow = sheet.getCell(matchingRowNum.value - 1, 0)
                    .getEntireRow().getUsedRange();
                studentRow.load("values");
            })
            .then(context.sync)
            .then(function() {
                // Hide all rows except header ones and the student row
                var cellB2AndOnward = sheet.getUsedRange()
                    .getOffsetRange(1, 1);
                cellB2AndOnward.rowHidden = true
                cellB2AndOnward.columnHidden = true;
                studentRow.rowHidden = false;

                for (var c = 0; c < studentRow.values[0].length; c++) {
                    if (studentRow.values[0][c] < 80) {
                        studentRow.getColumn(c).columnHidden = false;
                    }
                }

                studentRow.getCell(0, 0).select();
            })
            .then(context.sync);
    }).catch(OfficeHelpers.Utilities.log);
}

Samples/Chapter 5 - Core Concepts/Gradebook/Filter student grades (TypeScript).ts
function filterStudentGrades() {
    Excel.run(async (context) => {
        // #1: Find the row that matches the name of the student:

        let sheet = context.workbook.worksheets.getActiveWorksheet();
        let nameColumn = sheet.getRange("A:A");
        
        let studentName = $("#student-name").val();
        let matchingRowNum = context.workbook.functions.match(
            studentName, nameColumn, 0 /*exact match*/);
        matchingRowNum.load("value");
        
        await context.sync()


        // #2: Load the cell values (filtered to just the
        //     used range, to minimize data-transfer)

        let studentRow = sheet.getCell(matchingRowNum.value - 1, 0)
            .getEntireRow().getUsedRange().load("values");

        await context.sync();

        // Hide all cells except the header ones and the student row
        let cellB2AndOnward = sheet.getUsedRange()
            .getOffsetRange(1, 1).getResizedRange(-1, -1);
        cellB2AndOnward.rowHidden = true
        cellB2AndOnward.columnHidden = true;
        studentRow.rowHidden = false;

        // Turn the visiiblity back on for columns with low grades
        for (let c = 1; c < studentRow.values[0].length; c++) {
            if (studentRow.values[0][c] < 80) {
                studentRow.getColumn(c).columnHidden = false;
            }
        }

        studentRow.getCell(0, 0).select();

        await context.sync();

    }).catch(OfficeHelpers.Utilities.log);
}

Samples/Chapter 5 - Core Concepts/Canonical samples - population analysis (JavaScript).js
function analyzePopulationGrowth() {
    Excel.run(function (context) {
        // Create proxy objects to represent entities that are
        // in the actual workbook. More information on proxy objects 
        // will be presented in the very next section of this chapter.

        var originalTable = context.workbook.tables
            .getItem("PopulationTable");

        var nameColumn = originalTable.columns.getItem("City");
        var latestDataColumn = originalTable.columns.getItem(
            "7/1/2014 population estimate");
        var earliestDataColumn = originalTable.columns.getItem(
            "4/1/1990 census population");

        // Now, queue up a command to load the values for each
        // of the columns we want to read from.  Note that the
        // actual fetching and returning of the values is
        // deferred until a "context.sync()".

        nameColumn.load("values");
        latestDataColumn.load("values");
        earliestDataColumn.load("values");

        return context.sync()
            .then(function () {
                // Create an in-memory data representation, using an
                // array with JSON objects representing each city.
                var citiesData = [];

                // Start at i = 1 (that is, 2nd row of the table --
                // remember the 0-indexing) to skip the header.
                for (var i = 1; i < nameColumn.values.length; i++) {
                    var name = nameColumn.values[i][0];

                    // Note that because "values" is a 2D array
                    // (even if, in this case, it's just a single 
                    //  column), extract the 0th element of each row.
                    var pop1990 = earliestDataColumn.values[i][0];
                    var popLatest = latestDataColumn.values[i][0];

                    // A couple of the cities don't have data for 
                    // 1990,so skip over those.
                    if (isNaN(pop1990) || isNaN(popLatest)) {
                        console.log('Skipping "' + name + '"');
                    }

                    var growth = popLatest - pop1990;
                    citiesData.push({name: name, growth: growth});
                }

                var sorted = citiesData.sort(function (city1, city2) {
                    return city2.growth - city1.growth;
                    // Note the opposite order from the usual
                    // "first minus second" -- because want to sort in
                    // descending order rather than ascending.
                });
                var top10 = sorted.slice(0, 10);

                // Now that we've computed the data, create a new 
                // worksheet for the output, and write in the data:
                var sheetTitle = "Top 10 Growing Cities";
                var sheetHeaderTitle = "Population Growth 1990 - 2014";
                var tableCategories = ["Rank", "City", "Population Growth"];
                var outputSheet = context.workbook.worksheets.add(sheetTitle);

                var reportStartCell = outputSheet.getRange("B2");
                reportStartCell.values = [[sheetHeaderTitle]];
                reportStartCell.format.font.bold = true;
                reportStartCell.format.font.size = 14;
                reportStartCell.getResizedRange
                    (0, tableCategories.length - 1).merge();

                var tableHeader = reportStartCell.getOffsetRange(2, 0)
                    .getResizedRange(0, tableCategories.length - 1);
                tableHeader.values = [tableCategories];
                var table = outputSheet.tables.add(
                    tableHeader, true /*hasHeaders*/);

                for (var i = 0; i < top10.length; i++) {
                    var cityData = top10[i];
                    table.rows.add(
                        null /* null means "add to end" */,
                        [[i + 1, cityData.name, cityData.growth]]);

                    // Note: even though adding just a single row,
                    // the API still expects a 2D array (for 
                    // consistency and with Range.values)
                }

                // Auto-fit the column widths, and set uniform 
                // thousands-separator number-formatting on the
                // "Population" column of the table.
                table.getRange().getEntireColumn().format.autofitColumns();
                table.getDataBodyRange().getLastColumn()
                    .numberFormat = [["#,##"]];


                // Finally, with the table in place, add a chart:
                var fullTableRange = table.getRange();

                // For the chart, no need to show the "Rank", so 
                // only use the column with the city's name -- and
                // then expand it one column to the right
                // to include the population data as well.
                var dataRangeForChart = fullTableRange
                    .getColumn(1).getResizedRange(0, 1);

                var chart = outputSheet.charts.add(
                    Excel.ChartType.columnClustered,
                    dataRangeForChart,
                    Excel.ChartSeriesBy.columns);

                chart.title.text =
                    "Population Growth between 1990 and 2014";

                // Position the chart to start below the table,
                // occupy the full table width, and be 15 rows tall
                var chartPositionStart = fullTableRange
                    .getLastRow().getOffsetRange(2, 0);
                chart.setPosition(chartPositionStart,
                    chartPositionStart.getOffsetRange(14, 0));

                outputSheet.activate();
            })
            .then(context.sync);

    }).catch(function (error) {
        console.log(error);
        // Log additional debug information, if applicable:
        if (error instanceof OfficeExtension.Error) {
            console.log(error.debugInfo);
        }
    });
}


Samples/Chapter 5 - Core Concepts/Canonical samples - population analysis (TypeScript).ts
function analyzePopulationGrowth() {
    Excel.run(async (context) => {
        // Create proxy objects to represent the "real" workbook objects
        // that we'll be working with.  More information on proxy objects 
        // will be presented in the very next section of this chapter.

        let originalTable = context.workbook.tables
            .getItem("PopulationTable");

        let nameColumn = originalTable.columns.getItem("City");
        let latestPopulationColumn = originalTable.columns.getItem(
            "7/1/2014 population estimate");
        let earliestCensusColumn = originalTable.columns.getItem(
            "4/1/1990 census population");

        // Now, load the values for each of the three columns that we
        // want to read from.  Note that, to support batching operations
        // together (again, you'll see more in the upcoming sections
        // of this chapter), the load doesn't *actually* happen until
        // we do a "context.sync()", as below.

        nameColumn.load("values");
        latestPopulationColumn.load("values");
        earliestCensusColumn.load("values");

        await context.sync();


        // Create an in-memory representation of the data, using an 
        // array that will contain JSON objects representing each city.
        let citiesData: Array<{name: string, growth: number}> = [];

        // Start at i = 1 (that is, 2nd row of the table --
        // remember the 0-indexing) in order to skip the header.
        for (let i = 1; i < nameColumn.values.length; i++) {
            let name = nameColumn.values[i][0];

            // Note that because the "values" is a 2D array (even if,
            // in this particular case, it's just a single column),
            // need to extract out the 0th element of each row.
            let pop1990 = earliestCensusColumn.values[i][0];
            let popLatest = latestPopulationColumn.values[i][0];

            // A couple of the cities don't have data for 1990,
            // so skip over those.
            if (isNaN(pop1990) || isNaN(popLatest)) {
                console.log('Skipping "' + name + '"');
            }

            let growth = popLatest - pop1990;
            citiesData.push({name: name, growth: growth});
        }

        let sorted = citiesData.sort((city1, city2) => {
            return city2.growth - city1.growth;
            // Note the opposite order from the usual
            // "first minus second" -- because want to sort in
            // descending order rather than ascending.
        });
        let top10 = sorted.slice(0, 10);

        // Now that we've computed the data, create a new worksheet
        // for the output, and write in the data:
        let sheetTitle = "Top 10 Growing Cities";
        let sheetHeaderTitle = "Population Growth 1990 - 2014"; 
        let tableCategories = ["Rank", "City", "Population Growth"];
        let outputSheet = context.workbook.worksheets.add(sheetTitle);

        let reportStartCell = outputSheet.getRange("B2");
        reportStartCell.values = [[sheetHeaderTitle]];
        reportStartCell.format.font.bold = true;
        reportStartCell.format.font.size = 14;
        reportStartCell.getResizedRange
            (0, tableCategories.length - 1).merge();

        let tableHeader = reportStartCell.getOffsetRange(2, 0)
            .getResizedRange(0, tableCategories.length - 1);
        tableHeader.values = [ tableCategories ];
        let table = outputSheet.tables.add(
            tableHeader, true /*hasHeaders*/);

        for (let i = 0; i < top10.length; i++) {
            let cityData = top10[i];
            table.rows.add(
                null /* null means "add to end" */,
                [[i + 1, cityData.name, cityData.growth]]);

            // Note: even though adding just a single row, the API
            // still expects a 2D array for consistency and
            // interoperability with Range.values.
        }

        // Auto-fit the column widths, and set uniform thousands-separator
        // number formatting on the "Population" column of the table.
        table.getRange().getEntireColumn().format.autofitColumns();
        table.getDataBodyRange().getLastColumn()
            .numberFormat = [["#,##"]];


        // Finally, with the table in place, add a chart:
        let fullTableRange = table.getRange();

        // For the chart, no need to show the "Rank", so only use the
        //     column with the city's name, and expand it one column
        //     to the right to include the population data as well.
        let dataRangeForChart = fullTableRange
            .getColumn(1).getResizedRange(0, 1);

        let chart = outputSheet.charts.add(
            Excel.ChartType.columnClustered,
            dataRangeForChart,
            Excel.ChartSeriesBy.columns);

        chart.title.text = "Population Growth between 1990 and 2014";

        // Position the chart to start below the table,
        // occupy the full table width, and be 15 rows tall
        let chartPositionStart = fullTableRange
            .getLastRow().getOffsetRange(2, 0);
        chart.setPosition(chartPositionStart,
            chartPositionStart.getOffsetRange(14, 0));

        outputSheet.activate();

        await context.sync();

    }).catch((error) => {
        console.log(error);
        // Log additional information, if applicable:
        if (error instanceof OfficeExtension.Error) {
            console.log(error.debugInfo);
        }
    });
}


Samples/Chapter 5 - Core Concepts/Canonical samples - population analysis (JavaScript).ts
function analyzePopulationGrowth() {
    Excel.run(function (context) {
        // Create proxy objects to represent the "real" workbook objects
        // that we'll be working with.  More information on proxy objects 
        // will be presented in the very next section of this chapter.

        var originalTable = context.workbook.tables
            .getItem("PopulationTable");

        var nameColumn = originalTable.columns.getItem("City");
        var latestPopulationColumn = originalTable.columns.getItem(
            "7/1/2014 population estimate");
        var earliestCensusColumn = originalTable.columns.getItem(
            "4/1/1990 census population");

        // Now, load the values for each of the three columns that we
        // want to read from.  Note that, to support batching operations
        // together (again, you'll see more in the upcoming sections
        // of this chapter), the load doesn't *actually* happen until
        // we do a "context.sync()", as below.

        nameColumn.load("values");
        latestPopulationColumn.load("values");
        earliestCensusColumn.load("values");

        return context.sync()
            .then(function () {
                // Create an in-memory representation of the data, using an 
                // array that will contain JSON objects representing each city.
                var citiesData = [];

                // Start at i = 1 (that is, 2nd row of the table --
                // remember the 0-indexing) in order to skip the header.
                for (var i = 1; i < nameColumn.values.length; i++) {
                    var name = nameColumn.values[i][0];

                    // Note that because the "values" is a 2D array (even if,
                    // in this particular case, it's just a single column),
                    // need to extract out the 0th element of each row.
                    var pop1990 = earliestCensusColumn.values[i][0];
                    var popLatest = latestPopulationColumn.values[i][0];

                    // A couple of the cities don't have data for 1990,
                    // so skip over those.
                    if (isNaN(pop1990) || isNaN(popLatest)) {
                        console.log('Skipping "' + name + '"');
                    }

                    var growth = popLatest - pop1990;
                    citiesData.push({name: name, growth: growth});
                }

                var sorted = citiesData.sort(function (city1, city2) {
                    return city2.growth - city1.growth;
                    // Note the opposite order from the usual
                    // "first minus second" -- because want to sort in
                    // descending order rather than ascending.
                });
                var top10 = sorted.slice(0, 10);

                // Now that we've computed the data, create a new worksheet
                // for the output, and write in the data:
                var sheetTitle = "Top 10 Growing Cities";
                var sheetHeaderTitle = "Population Growth 1990 - 2014";
                var tableCategories = ["Rank", "City", "Population Growth"];
                var outputSheet = context.workbook.worksheets.add(sheetTitle);

                var reportStartCell = outputSheet.getRange("B2");
                reportStartCell.values = [[sheetHeaderTitle]];
                reportStartCell.format.font.bold = true;
                reportStartCell.format.font.size = 14;
                reportStartCell.getResizedRange
                    (0, tableCategories.length - 1).merge();

                var tableHeader = reportStartCell.getOffsetRange(2, 0)
                    .getResizedRange(0, tableCategories.length - 1);
                tableHeader.values = [tableCategories];
                var table = outputSheet.tables.add(
                    tableHeader, true /*hasHeaders*/);

                for (var i = 0; i < top10.length; i++) {
                    var cityData = top10[i];
                    table.rows.add(
                        null /* null means "add to end" */,
                        [[i + 1, cityData.name, cityData.growth]]);

                    // Note: even though adding just a single row, the API
                    // still expects a 2D array for consistency and
                    // interoperability with Range.values.
                }

                // Auto-fit the column widths, and set uniform thousands-separator
                // number formatting on the "Population" column of the table.
                table.getRange().getEntireColumn().format.autofitColumns();
                table.getDataBodyRange().getLastColumn()
                    .numberFormat = [["#,##"]];


                // Finally, with the table in place, add a chart:
                var fullTableRange = table.getRange();

                // For the chart, no need to show the "Rank", so only use the
                //     column with the city's name, and expand it one column
                //     to the right to include the population data as well.
                var dataRangeForChart = fullTableRange
                    .getColumn(1).getResizedRange(0, 1);

                var chart = outputSheet.charts.add(
                    Excel.ChartType.columnClustered,
                    dataRangeForChart,
                    Excel.ChartSeriesBy.columns);

                chart.title.text = "Population Growth between 1990 and 2014";

                // Position the chart to start below the table,
                // occupy the full table width, and be 15 rows tall
                var chartPositionStart = fullTableRange
                    .getLastRow().getOffsetRange(2, 0);
                chart.setPosition(chartPositionStart,
                    chartPositionStart.getOffsetRange(14, 0));

                outputSheet.activate();
            })
            .then(context.sync);

    }).catch(function (error) {
        console.log(error);
        // Log additional information, if applicable:
        if (error instanceof OfficeExtension.Error) {
            console.log(error.debugInfo);
        }
    });
}
